src/VectorBundle.jl

```diff
@@ -0,0 +1,372 @@
+end
+
+function zero_vector!(M::VectorBundleFibers{<:TangentSpaceType}, v, x)
+    zero_tangent_vector!(M.M, v, x)
+    return v
+end
+
+"""
+    zero_vector(M::VectorBundleFibers, x)
+
+Compute the zero vector from the vector space of type `VS` at point `x`
+from manifold `M.M`.
+"""
+function zero_vector(M::VectorBundleFibers, x)
+    v = similar_result(M, zero_vector, x)
+    zero_vector!(M, v, x)
+    return v
+end
+
+"""
+    project_vector!(M::VectorBundleFibers, v, x, w)
+
+Project vector `w` from the vector space of type `VS` at point `x`
+and save the result to `v`.
+"""
+function project_vector!(M::VectorBundleFibers, v, x, w)
+    error("project_vector! not implemented for vector space manifold $(typeof(M)), vector space of type $(typeof(VS)), output vector of type $(typeof(v)) and input vector at point $(typeof(x)) with type of w $(typeof(w)).")
+end
+
+function project_vector!(M::VectorBundleFibers{<:TangentSpaceType}, v, x, w)
+    project_tangent!(M.M, v, x, w)
+    return v
+end
+
+"""
+    VectorBundle(M::Manifold, type::VectorSpaceType)
+
+Vector bundle on manifold `M` of type `type`.
+"""
+struct VectorBundle{TVS<:VectorSpaceType, TM<:Manifold} <: Manifold
+    type::TVS
+    M::TM
+    VS::VectorBundleFibers{TVS, TM}
+end
+
+function VectorBundle(VS::TVS, M::TM) where{TVS<:VectorSpaceType, TM<:Manifold}
+    return VectorBundle{TVS, TM}(VS, M, VectorBundleFibers(VS, M))
+end
+
+function representation_size(M::VectorBundle)
+    len_manifold = prod(representation_size(M.M))
+    len_vs = prod(representation_size(M.VS))
+    return (len_manifold + len_vs,)
+end
+
+TangentBundleFibers{M} = VectorBundleFibers{TangentSpaceType,M}
+TangentBundleFibers(M::Manifold) = VectorBundleFibers(TangentSpace, M)
+
+CotangentBundleFibers{M} = VectorBundleFibers{CotangentSpaceType,M}
+CotangentBundleFibers(M::Manifold) = VectorBundleFibers(CotangentSpace, M)
+
+TangentBundle{M} = VectorBundle{TangentSpaceType,M}
+TangentBundle(M::Manifold) = VectorBundle(TangentSpace, M)
+
+CotangentBundle{M} = VectorBundle{CotangentSpaceType,M}
+CotangentBundle(M::Manifold) = VectorBundle(CotangentSpace, M)
+
+"""
+    bundle_projection(M::VectorBundle, x::ProductRepr)
+
+Projection of point `x` from the bundle `M` to the base manifold.
+Returns the point on the base manifold `M.M` at which the vector part
+of `x` is attached.
+"""
+function bundle_projection(M::VectorBundle, x)
+    return submanifold_component(x, 1)
+end
+
+function isapprox(M::VectorBundle, x, y; kwargs...)
+    return isapprox(M.M, x.parts[1], y.parts[1]; kwargs...) &&
+        isapprox(x.parts[2], y.parts[2]; kwargs...)
+end
+
+function isapprox(M::VectorBundle, x, v, w; kwargs...)
+    return isapprox(M.M, v.parts[1], w.parts[1]; kwargs...) &&
+        isapprox(M.M, x.parts[1], v.parts[2], w.parts[2]; kwargs...)
+end
+
+manifold_dimension(M::VectorBundle) = manifold_dimension(M.M) + vector_space_dimension(M.VS)
+
+function inner(M::VectorBundle, x, v, w)
```

Comment:
I don't understand what this function means. What is an inner product on a tangent bundle, for example? What is `x`, and why does it have parts?

Reply:
Oh, I see. `x` is a point with a tangent vector. I don't really have a concept for the tangent space to a vector bundle. Perhaps documentation could explain `v` and `w`?

Reply:
Vector bundles are manifolds, so they can have vector bundles themselves You can look here: https://en.wikipedia.org/wiki/Double_tangent_bundle . I'll make a better general documentation for vector bundles.

Reply:
🤯